File: benchmarks/call-closure-poly4/bench.js
function makeAdder(base) {
  return function (x) {
    return x + base;
  };
}

function main() {
  const iterations = 20000000;
  const f0 = makeAdder(1);
  const f1 = makeAdder(3);
  const f2 = makeAdder(5);
  const f3 = makeAdder(7);
  let acc = 0;

  for (let i = 0; i < iterations; i++) {
    const x = i & 1023;
    const slot = i & 3;
    if (slot === 0) {
      acc += f0(x);
    } else if (slot === 1) {
      acc += f1(x);
    } else if (slot === 2) {
      acc += f2(x);
    } else {
      acc += f3(x);
    }
  }

  console.log(acc);
}

main();

File: benchmarks/call-closure-poly4/bench.lua
local function makeAdder(base)
  return function(x)
    return x + base
  end
end

local iterations = 20000000
local f0 = makeAdder(1)
local f1 = makeAdder(3)
local f2 = makeAdder(5)
local f3 = makeAdder(7)
local acc = 0

for i = 0, iterations - 1 do
  local x = i % 1024
  local slot = i % 4
  if slot == 0 then
    acc = acc + f0(x)
  elseif slot == 1 then
    acc = acc + f1(x)
  elseif slot == 2 then
    acc = acc + f2(x)
  else
    acc = acc + f3(x)
  end
end

print(acc)
